### PR TITLE
Add paginated collectibles

### LINKED ISSUE #920: Add paginated endpoint for collectibles
**Is your feature request related to a problem? Please describe.**
Long lists of collectibles ERC720 takes a lot of time to return to the client.   
 
- https://github.com/safe-global/safe-pm/issues/89

**Describe the solution you'd like**
Include a new version of endpoint to support pagination of transaction service   
- https://github.com/safe-global/safe-transaction-service/issues/239

### PR DESCRIPTION
**Pull request by moisses89**
Closes #920 
- add paginated collectibles 
- add test for paginated collectibles

### CODE REVIEW

File: src/routes/collectibles/routes.rs [hunk 1]
```diff
@@ -165,3 +170,45 @@ pub async fn get_collectibles(
...
+#[get("/v2/chains/<chain_id>/safes/<safe_address>/collectibles?<cursor>&<trusted>&<exclude_spam>")]
+pub async fn get_collectibles_paginated(
+    context: RequestContext,
+    chain_id: String,
+    safe_address: String,
+    cursor: Option<String>,
+    trusted: Option<bool>,
+    exclude_spam: Option<bool>,
+) -> ApiResult<content::RawJson<String>> {
+    CacheResponse::new(&context, ChainCache::Other)
```
Comment (moisses89) [inline src/routes/collectibles/routes.rs:200]: I will review this, ChainCache?

Comment (moisses89) [reply src/routes/collectibles/routes.rs:200]: @fmrsabino what do you think? would be necessary caching the response?

Comment (Uxio0) [reply src/routes/collectibles/routes.rs:200]: I think we should always cache in the cgw, as in the other endpoints

File: src/routes/collectibles/handlers.rs [hunk 2]
```diff
@@ -30,3 +36,76 @@ pub async fn collectibles(
...
+        .map(|cursor| PageMetadata::from_cursor(cursor));
+
+    let url = core_uri!(
+        info_provider,
+        "/v2/safes/{}/collectibles/?{}&trusted={}&exclude_spam={}", /* paginated service in core is behind v2 api. */
+        safe_address,
+        page_metadata
+            .as_ref()
+            .unwrap_or(&PageMetadata::default())
+            .to_url_string(),
```
Comment (fmrsabino) [inline src/routes/collectibles/handlers.rs:61]: nitpick: I'd move this up where you map from the `cursor` (lines 50 to 52) since we don't need the `page_metadata` for anything else

File: src/routes/collectibles/handlers.rs [hunk 3]
```diff
@@ -30,3 +36,76 @@ pub async fn collectibles(
...
+    exclude_spam: Option<bool>,
+) -> ApiResult<Json<Page<ServiceCollectible>>> {
+    let info_provider = DefaultInfoProvider::new(chain_id, &context);
+    let page_metadata = cursor
+        .as_ref()
+        .map(|cursor| PageMetadata::from_cursor(cursor));
+
+    let url = core_uri!(
+        info_provider,
+        "/v2/safes/{}/collectibles/?{}&trusted={}&exclude_spam={}", /* paginated service in core is behind v2 api. */
```
Comment (fmrsabino) [inline src/routes/collectibles/handlers.rs:56]: I would remove this comment since this should be self-explanatory

File: src/routes/collectibles/handlers.rs [hunk 4]
```diff
@@ -30,3 +36,76 @@ pub async fn collectibles(
...
+    original_link: Option<String>,
+    chain_id: String,
+    safe_address: String,
+    trusted: Option<bool>,
+    exclude_spam: Option<bool>,
+) -> Option<String> {
+    original_link.as_ref().map(|link| {
+        let cursor =
+            PageMetadata::from_cursor(link.split("?").collect::<Vec<&str>>().get(1).unwrap_or(&""))
+                .to_url_string();
```
Comment (fmrsabino) [inline src/routes/collectibles/handlers.rs:96]: For the context of this PR this is ok but we need to think of ways to refactor this. This logic is spread across multiple places.

Also URL parsing/manipulation should probably be `Url` which we do have in the project.

File: src/tests/json/collectibles/collectibles_page.json [hunk 5]
```diff
@@ -108,3 +108,4 @@
         }
     }
 ]
+
```
Comment (fmrsabino) [inline src/tests/json/collectibles/collectibles_page.json:111]: We should revert this change (double new line at the end)

File: src/routes/collectibles/handlers.rs [hunk 6]
```diff
@@ -1,10 +1,16 @@
+use std::cmp::max;
```
Comment (fmrsabino) [inline src/routes/collectibles/handlers.rs:1]: This is not required right?

File: src/routes/collectibles/routes.rs [hunk 7]
```diff
@@ -1,8 +1,16 @@
+use crate::cache::cache_operations::CacheResponse;
+use crate::cache::manager::ChainCache;
```
Comment (fmrsabino) [inline src/routes/collectibles/routes.rs:2]: These should also not be needed

File: src/routes/collectibles/routes.rs [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
+use crate::common::models::page::Page;
+use crate::config::chain_info_response_cache_duration;
```
Comment (fmrsabino) [inline src/routes/collectibles/routes.rs:4]: This one should also not be needed.

**Review by fmrsabino — CHANGES_REQUESTED** [review]
Looks good 👏 

Just some small requests.

File: src/routes/collectibles/routes.rs (hunk 1 above)
Comment (fmrsabino) [reply src/routes/collectibles/routes.rs]: > I think we should always cache in the cgw, as in the other endpoints

That's correct. However for the collectibles, the first version of the endpoint only cached the request to the Safe Transaction Service and not the response given by the Safe Client Gateway to the clients.

> @fmrsabino what do you think? would be necessary caching the response?

I would say that we can start by not caching the response to the clients. We might be introducing unnecessary complexity. I'd roll out this endpoint and keep track of its usage.

Comment (moisses89) [reply src/routes/collectibles/routes.rs]: Ok, I agree with only chaching the request.

File: src/routes/collectibles/handlers.rs (hunk 6 above)
Comment (moisses89) [reply src/routes/collectibles/handlers.rs:1]: yes, It's not required

File: src/routes/collectibles/handlers.rs (hunk 4 above)
Comment (moisses89) [reply src/routes/collectibles/handlers.rs:96]: I agree, what do you think to create an issue to improve the manage of pagination and refactoring the source?

File: src/routes/collectibles/handlers.rs (hunk 2 above)
Comment (moisses89) [reply src/routes/collectibles/handlers.rs]: Clearest

File: src/routes/collectibles/handlers.rs (hunk 4 above)
Comment (fmrsabino) [reply src/routes/collectibles/handlers.rs:94]: Yes I think we should create one. No need to specify the solution (just that it needs to be evaluated)

**Review by fmrsabino — APPROVED** [review]
